src/typechecker.hrl

```diff
@@ -1,13 +1,17 @@
 -ifndef(__TYPECHECKER_HRL__).
 -define(__TYPECHECKER_HRL__, true).
 
+-record(clause_block, {exhaust}).
+
 -record(env, {fenv              = #{}   :: map(),
               imported          = #{}   :: #{{atom(), arity()} => module()},
               venv              = #{}   :: typechecker:venv(),
               tenv                      :: gradualizer_lib:tenv(),
               infer             = false :: boolean(),
               verbose           = false :: boolean(),
               exhaust           = true  :: boolean(),
+              %% Options tunable per clause block (fun/case/try-catch/receive).
+              blocks            = []    :: [#clause_block{}],
```

Comment:
I think most programming languages have the concept of a code block, e.g. in C (almost?) everywhere where we can use a single expression we can also use a `{ ... }` delimited block.

> Maybe clause_stack is a better name here?

I was thinking of that and also wondering about `clauses_stack`, but ultimately decided not to make it more complicated than necessary. `clause_stack` (singular clause) is not correct, because the stack element is pushed/popped not per clause, but per a group of all clauses making up the higher-level expression (fun, case, try ... catch, or receive) - I think it would be misleading. I was also considering a `form_stack`, but that's not correct either, since we actually have multiple levels in a top-level form, and in general forms are more general than clauses or clause blocks.

The type checker code uses `Clauses` (plural) to name the concept for which we push/pop the things. However, using a plural noun to describe other nouns depending on it makes room for ambiguity (does the description refer to a single instance or to the grouping?). I think that "code block" might not be perfect, but IMO is good enough :)

Reply:
BTW, I'll fix the comment, you're right it's a bit misleading.

Reply:
In C and Elixir, a block is always `{...}` and `do-end` respectively (regardless of whether it's a list of statements `{A; B; C;}` or a `while (...) {...}` or `switch {...}`), so it makes some sense.

In C, you can also have a sequence of expressions, delimited by comma as in Erlang, but that's not a block since there are no braces. Example: `(i = 0, j = 42, i + j)`.

In Erlang, the're so do-end; there's different syntax for functions, expressions, clauses, etc. More importantly, the word "block" is not used to describe anything in Erlang grammar in the official docs.

I think `clauses_stack` is perfect, especially if "clauses" is used in the grammar docs. :-)

Reply:
Ok, good point that it's used like that in the grammar. I'll change it :+1:

Reply:
Fixed, WDYT?